File: HASH.DiscountCalculator/HASH.DiscountCalculator/Models/Product.cs
using MongoDB.Bson.Serialization.Attributes;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace HASH.DiscountCalculator.Models
{
    public class Product
    {
        public Product()
        {
            Discount = new Discount();
        }

        [BsonId]
        [BsonRepresentation(MongoDB.Bson.BsonType.ObjectId)]
        public string Id { get; set; }
        public int PriceCents { get; set; }
        public string Title { get; set; }
        public string Description { get; set; }
        public Discount Discount { get; set; }

        public void CheckBirthDayDiscount(DateTime userBirthDay)
        {
            if (userBirthDay.Date == DateTime.Now.Date) {
                Discount.Percentage = 0.05F;
                Discount.ValueCents = (int)(PriceCents * Discount.Percentage);              
            }

            CheckDiscountLimit();
        }

        public void CheckBlackFridayDiscount(DateTime now)
        {
            if (now.Day == Dates.BlackFridayDay && now.Month == Dates.BlackFridayMonth)
            {
                Discount.Percentage = 0.1F;
                Discount.ValueCents = (int)(PriceCents * Discount.Percentage);
            }

            CheckDiscountLimit();
        }

        public void CheckDiscountLimit()
        {
            if (Discount.Percentage > 0.1F)
            {
                Discount.Percentage = 0;
                Discount.ValueCents = 0;
            }
        }

        public ProductModel ParseToProductModel()
        {
            return new ProductModel()
            {
                Id = Id,
                PriceInCents = PriceCents,
                Title = Title,
                Description = Description,
                Discount = new DiscountModel()
                {
                    Pct = Discount.Percentage,
                    ValueInCents = Discount.ValueCents
                }
            };
        }
    }
}

File: HASH.ProductsList/server/client/js/app/models/Product.js
class Product{

    constructor(id, price_in_cents, title, description, discount){
        
        this._id = id;
        this._price_in_cents = price_in_cents;
        this._title = title;
        this._description = description;
        this._discount = discount;
        
    } 

    get id() {
        
        return this._id;
    }

    get price_in_cents() {
        
        return this._price_in_cents;
    }

    get title() {
        
        return this._title;
    }

    get description() {
        
        return this._description;
    }

    get discount() {
        
        return this._discount;
    }

    setDiscount(discount){
        return this._discount = discount;
    }

}